autoload/lsp/buffer.vim

```diff
@@ -41,8 +41,35 @@ export def BufLspServerGet(bnr: number): dict<any>
     return {}
   endif
 
-  # TODO implement logic to compute which server to return
-  return bufnrToServers[bnr][0]
+  if domain == null_string
+    return bufnrToServers[bnr][0]
+  endif
+
+  var SupportedCheckFns = {
+  }
+
+  if !SupportedCheckFns->has_key(domain)
+    # If this happns it is a programming error, and should be fixed in the source code
+    :throw $'Error: ''{domain}'' is not a valid domain'
+    return {}
+  endif
+
+  var SupportedCheckFn = SupportedCheckFns[domain]
+
+  for lspserver in bufnrToServers[bnr]
+    if !SupportedCheckFn(lspserver)
+      continue
+    endif
+
+    if !lspserver.customProviding->has_key(domain)
+        || !lspserver.customProviding[domain](lspserver)
+      continue
+    endif
+
+    return lspserver
+  endfor
+
+  return {}
```

Comment:
The idea behind the code is to use the first server that actually supports a given action. But also letting the user overriding which language server which should be used by providing a: `customProving` (⚠️ needs a better name) dictionary with functions that returns boolean if a language server should support a action.